x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/rules_management_client.ts

```diff
@@ -0,0 +1,379 @@
+      return deleteRule(rulesClient, deleteRulePayload);
+    },
+
+    upgradePrebuiltRule: async (
+      upgradePrebuiltRulePayload: UpgradePrebuiltRuleProps
+    ): Promise<RuleAlertType> => {
+      return upgradePrebuiltRule(rulesClient, upgradePrebuiltRulePayload);
+    },
+
+    importRule: async (importRulePayload: ImportRuleProps): Promise<RuleAlertType> => {
+      return importRule(rulesClient, importRulePayload);
+    },
+  };
+
+  return client;
+};
+
+export const createCustomRule = async (
+  rulesClient: RulesClient,
+  createCustomRulePayload: CreateCustomRuleProps
+): Promise<RuleAlertType> => {
+  const { params } = createCustomRulePayload;
+
+  const rule = await _createRule(rulesClient, params, { immutable: false });
+  return rule;
+};
+
+export const createPrebuiltRule = async (
+  rulesClient: RulesClient,
+  createPrebuiltRulePayload: CreatePrebuiltRuleProps
+): Promise<RuleAlertType> => {
+  const { ruleAsset } = createPrebuiltRulePayload;
+
+  const rule = await _createRule(rulesClient, ruleAsset, {
+    immutable: true,
+    defaultEnabled: false,
+  });
+
+  return rule;
+};
+
+export const updateRule = async (
+  rulesClient: RulesClient,
+  updateRulePayload: UpdateRuleProps
+): Promise<RuleAlertType> => {
+  const { ruleUpdate, existingRule } = updateRulePayload;
+
+  const update = await _updateRule(rulesClient, { ruleUpdate, existingRule });
+
+  const enabled = ruleUpdate.enabled ?? existingRule.enabled;
+  await _toggleRuleEnabledOnUpdate(rulesClient, existingRule, enabled);
+
+  return { ...update, enabled };
+};
+
+export const patchRule = async (
+  rulesClient: RulesClient,
+  patchRulePayload: PatchRuleProps
+): Promise<RuleAlertType> => {
+  const { nextParams, existingRule } = patchRulePayload;
+  const update = await _patchRule(rulesClient, patchRulePayload);
+
+  await _toggleRuleEnabledOnUpdate(rulesClient, existingRule, nextParams.enabled ?? false);
+
+  if (nextParams.enabled != null) {
+    return { ...update, enabled: nextParams.enabled };
+  } else {
+    return update;
+  }
+};
+
+export const deleteRule = async (
+  rulesClient: RulesClient,
+  deleteRulePayload: DeleteRuleProps
+): Promise<void> => {
+  const { ruleId } = deleteRulePayload;
+  await rulesClient.delete({ id: ruleId });
+};
+
+export const upgradePrebuiltRule = async (
+  rulesClient: RulesClient,
+  upgradePrebuiltRulePayload: UpgradePrebuiltRuleProps
+): Promise<RuleAlertType> => {
+  const { ruleAsset } = upgradePrebuiltRulePayload;
+  const existingRule = await readRules({
+    rulesClient,
+    ruleId: ruleAsset.rule_id,
+    id: undefined,
+  });
+
+  if (!existingRule) {
+    throw new PrepackagedRulesError(`Failed to find rule ${ruleAsset.rule_id}`, 500);
+  }
+
+  // If rule has change its type during upgrade, delete and recreate it
+  if (ruleAsset.type !== existingRule.params.type) {
+    return _upgradePrebuiltRuleWithTypeChange(rulesClient, ruleAsset, existingRule);
```

Comment:
`_upgradePrebuiltRuleWithTypeChange` is used only here, an we inline it?

Reply:
While I agree that it's better to inline the `import` functions, here I prefer to keep the logic encapsualted within `_upgradePrebuiltRuleWithTypeChange`, since what that function does it not obvious. It needs some explanation via a comment block, and does two things, first deletes and then recreates. I also think that the function is big enough that it would bloat `upgradePrebuiltRule` if inlined and would make readibility even worse.